### PR TITLE
refactor rpc base framework

### PR DESCRIPTION
**Pull request by typhoonzero**
Changes:

- variable response do not hold scope, it use a callback to get variable, or hold the var ref.
- handlers have full control of what the request should be processed on server.
- move handlers to one folder.
- sync handlers and async handlers are different handlers
- `grpc_request.h` have only one class
- out_var do not use a pointer of pointer.
- A barrier class
- rpc_server only hold send_barrier and recv_barrier, do not hold one barrier for each rpc type
- rpc_server hold one barrier for each marked variable to mark variable ready state for get and barrier all workers.
- README for sync server scheme

- TODO: will update brpc folder in another PR

### CODE REVIEW

File: paddle/fluid/operators/distributed/handlers/get_handler.cc [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+    return true;
+  }
+  rpc_server_->WaitState(RPCServerState::STATE_RECV);
+  request->out_var_ = scope_->FindVar(request->varname_);
+  request->out_var_name_ = request->varname_;
+  return true;
+}
+
+framework::Variable* GetHandlerAsync::GetOrCreateRequestVar(
+    const std::string& varname, RPCRequest* request) {
```
Comment (seiriosPlus) [inline paddle/fluid/operators/distributed/handlers/get_handler.cc:42]: ```request ``` is not be used.

File: cmake/generic.cmake [hunk 2]
```diff
@@ -768,11 +768,11 @@ function(grpc_library TARGET_NAME)
     COMPILE_FLAGS  "-Wno-non-virtual-dtor -Wno-error=non-virtual-dtor -Wno-error=delete-non-virtual-dtor")
   cc_library("${TARGET_NAME}_grpc" SRCS "${grpc_grpc_srcs}")
 
-  set_source_files_properties(
-    ${grpc_library_SRCS}
-    PROPERTIES
-    COMPILE_FLAGS  "-Wno-non-virtual-dtor -Wno-error=non-virtual-dtor -Wno-error=delete-non-virtual-dtor")
-  cc_library("${TARGET_NAME}" SRCS "${grpc_library_SRCS}" DEPS "${TARGET_NAME}_grpc" "${TARGET_NAME}_proto" "${grpc_library_DEPS}")
+  # set_source_files_properties(
```
Comment (gongweibao) [inline cmake/generic.cmake]: Clean up?

File: paddle/fluid/operators/distributed/brpc/brpc_server.cc [hunk 3]
```diff
@@ -16,329 +16,149 @@
...
 
-    it = rpc_call_map.find(distributed::kRequestGetNoBarrier);
-    if (it != rpc_call_map.end()) {
-      request_getnobarrier_h_ = it->second;
-      getnobarrier_threads_.reset(new paddle::framework::ThreadPool(
-          rpc_server_->GetThreadNum(distributed::kRequestGetNoBarrier)));
+    it = rpc_call_map->find(distributed::RequestType::RECV_NO_BARRIER);
+    if (it != rpc_call_map->end()) {
+      getnobarrier_threads_.reset(new paddle::framework::ThreadPool(5));
+      // rpc_server_->GetThreadNum(distributed::kRequestGetNoBarrier)));
```
Comment (gongweibao) [inline paddle/fluid/operators/distributed/brpc/brpc_server.cc]: Why not use the registed threadnum?

File: paddle/fluid/operators/distributed/brpc/brpc_server.cc [hunk 4]
```diff
@@ -16,329 +16,149 @@
...
 
-    it = rpc_call_map.find(distributed::kRequestPrefetch);
-    if (it != rpc_call_map.end()) {
-      request_prefetch_h_ = it->second;
-      prefetch_threads_.reset(new paddle::framework::ThreadPool(
-          rpc_server_->GetThreadNum(distributed::kRequestPrefetch)));
+    it = rpc_call_map->find(distributed::RequestType::PREFETCH);
+    if (it != rpc_call_map->end()) {
+      prefetch_threads_.reset(new paddle::framework::ThreadPool(5));
+      // rpc_server_->GetThreadNum(distributed::kRequestPrefetch)));
```
Comment (gongweibao) [inline paddle/fluid/operators/distributed/brpc/brpc_server.cc]: Clean?

File: paddle/fluid/operators/distributed/brpc/brpc_server.cc [hunk 5]
```diff
@@ -16,329 +16,149 @@
...
-      request_get_monomer_handler_h_ = it->second;
-    }
-
-    it = rpc_call_map.find(distributed::kRequestGetMonomerBarrier);
-    if (it != rpc_call_map.end()) {
-      request_get_monomer_barrier_handler_h_ = it->second;
+    it = rpc_call_map->find(distributed::RequestType::CHECKPOINT);
+    if (it != rpc_call_map->end()) {
+      checkpoint_notify_threads_.reset(new paddle::framework::ThreadPool(5));
+      // rpc_server_->GetThreadNum(distributed::kRequestPrefetch)));
```
Comment (gongweibao) [inline paddle/fluid/operators/distributed/brpc/brpc_server.cc]: Clean?

**Review by gongweibao — CHANGES_REQUESTED** [review]

File: paddle/fluid/operators/distributed/brpc/brpc_server.cc (hunk 5 above)
Comment (typhoonzero) [reply paddle/fluid/operators/distributed/brpc/brpc_server.cc]: Done.

**Pull request closed without merging**